### PR TITLE
Public Cloud: Introduce registercloudguest test

### PR DESCRIPTION
**Pull request by pdostal**
There are two parts:
 * `publiccloud::instance::check_guestregister()` executed right after the instance boots up.
 * `publiccloud/registercloudguest` test module deregistering, reregistering and force registering the instance.
 * `publiccloud/register_system` ➡️ `publiccloud/register_addons` register only the addons **(new change)**.

This test follows [SUSE-Enceladus/cloud-regionsrv-client/integration_test-process.txt](https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/integration_test-process.txt)

- Related ticket: [poo#109304](https://progress.opensuse.org/issues/109304)
- Verification run: [12-SP4 GCE BYOS containers](http://pdostal-server.suse.cz/tests/14361), [12-SP5 EC2 img_proof](http://pdostal-server.suse.cz/tests/14359), [15-SP1 Azure BYOS gen2 consoletests](http://pdostal-server.suse.cz/tests/14360), [15-SP3 EC2 consoletests](http://pdostal-server.suse.cz/tests/14362), [15-SP3 EC2 BYOS ARM containers](http://pdostal-server.suse.cz/tests/14354), [15-SP4 Azure BYOS gen2 fio](http://pdostal-server.suse.cz/tests/14348)

### CODE REVIEW

**Comment by pdostal** [comment]
The pull request is *blocked* due to [bsc#1198080](https://bugzilla.suse.com/show_bug.cgi?id=1198080). Although please feel free to review.

File: lib/publiccloud/instance.pm [hunk 1]
```diff
@@ -203,10 +203,57 @@ sub wait_for_guestregister
...
+
+Check that the registration state of newly created guest match our requirements.
+https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/integration_test-process.txt
+=cut
+sub check_guestregister
+{
+    my ($self) = @_;
+
+    if (is_byos) {
+        $self->run_ssh_command(cmd => '! sudo zypper lr', timeout => 6000);
```
Comment (jlausuch) [inline lib/publiccloud/instance.pm:223]: why is this needed? If we want to make sure repos are not defined, why don't we use the same approach as below?
```
if ($self->run_ssh_command(cmd => '! sudo zypper lr', timeout => 6000, proceed_on_failure => 1, quiet => 1) !~ /no repos defined/) {
```
ofc, not that output message :)

File: lib/publiccloud/utils.pm [hunk 2]
```diff
@@ -79,15 +79,36 @@ sub register_addon {
     record_info('SUSEConnect time', 'The command SUSEConnect -r ' . get_addon_fullname($addon) . ' took ' . (time() - $cmd_time) . ' seconds.');
 }
 
+sub deregister_addon {
+    my ($remote, $addon) = @_;
+    my $arch = get_var('PUBLIC_CLOUD_ARCH') // "x86_64";
```
Comment (jlausuch) [inline lib/publiccloud/utils.pm:84]:
```suggestion
    my $arch = get_var('PUBLIC_CLOUD_ARCH', 'x86_64');
```

File: lib/publiccloud/utils.pm [hunk 3]
```diff
@@ -79,15 +79,36 @@ sub register_addon {
...
-    my $regcode = get_required_var('SCC_REGCODE');
+    my $regcode = (is_byos()) ? '-r ' . get_required_var('SCC_REGCODE') : '';
     my $remote = $instance->username . '@' . $instance->public_ip;
-    # not all images currently have registercloudguest pre-installed .
+    # not all BYOS images currently have registercloudguest pre-installed .
     # in such a case,we need to regsiter against SCC and install registercloudguest with all needed dependencies and then
     # unregister and re-register with registercloudguest
     if ($instance->run_ssh_command(cmd => "sudo which registercloudguest > /dev/null; echo \"registercloudguest\$?\" ", proceed_on_failure => 1) =~ m/registercloudguest1/) {
-        $instance->retry_ssh_command(cmd => "sudo SUSEConnect -r $regcode", timeout => 420, retry => 3, delay => 120);
+        $instance->retry_ssh_command(cmd => "sudo SUSEConnect $regcode", timeout => 420, retry => 3, delay => 120);
```
Comment (jlausuch) [inline lib/publiccloud/utils.pm:111]: if is on-demand image, it will execute only `sudo SUSEConnect`. What's the point?

File: lib/registration.pm [hunk 4]
```diff
@@ -223,14 +224,30 @@ sub ssh_add_suseconnect_product {
...
+wrapper for suseconnect -d $name.
 =cut
 sub remove_suseconnect_product {
     my ($name, $version, $arch, $params) = @_;
     $version //= scc_version();
-    $arch //= get_required_var('ARCH');
+    $arch //= get_required_var('arch');
     $params //= '';
-    script_retry("SUSEConnect -d -p $name/$version/$arch $params", retry => 5, delay => 60, timeout => 180);
+    script_retry("suseconnect -d -p $name/$version/$arch $params", retry => 5, delay => 60, timeout => 180);
```
Comment (jlausuch) [inline lib/registration.pm:234]: I've never seen this command in lower case, why do you change it?

File: lib/publiccloud/instance.pm (hunk 1 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:223]: In [SUSE-Enceladus/cloud-regionsrv-client/integration_test-process.txt](https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/integration_test-process.txt) is required both.

File: lib/publiccloud/utils.pm (hunk 3 above)
Comment (pdostal) [reply lib/publiccloud/utils.pm:111]: Good catch :wink:

File: lib/publiccloud/utils.pm [hunk 5]
```diff
@@ -79,15 +79,36 @@ sub register_addon {
...
+    } else {
+        ssh_remove_suseconnect_product($remote, get_addon_fullname($addon), undef, $arch, '', $timeout, $retries, $delay);
+    }
+    record_info('SUSEConnect time', 'The command SUSEConnect -d ' . get_addon_fullname($addon) . ' took ' . (time() - $cmd_time) . ' seconds.');
+}
+
 sub registercloudguest {
     my ($instance) = @_;
-    my $regcode = get_required_var('SCC_REGCODE');
+    my $regcode = (is_byos()) ? '-r ' . get_required_var('SCC_REGCODE') : '';
```
Comment (grisu48) [inline lib/publiccloud/utils.pm:105]: I would expect a variable `$regcode` to only contain the REGCODE and not an additional argument for a command that follows later. I would move the `-r` to `SUSEConnect` and `registercloudguest` do here simply

```suggestion
    my $regcode = (is_byos()) ? get_required_var('SCC_REGCODE') : '';
```

File: lib/registration.pm [hunk 6]
```diff
@@ -223,16 +224,32 @@ sub ssh_add_suseconnect_product {
...
     remove_suseconnect_product($name, [$version, [$arch, [$params]]]);
 
-Wrapper for SUSEConnect -d $name.
+wrapper for suseconnect -d $name.
 =cut
 sub remove_suseconnect_product {
     my ($name, $version, $arch, $params) = @_;
     $version //= scc_version();
-    $arch //= get_required_var('ARCH');
+    $arch //= get_required_var('arch');
```
Comment (grisu48) [inline lib/registration.pm:232]: Why is this now lowercase?

File: tests/publiccloud/registercloudguest.pm [hunk 7]
```diff
@@ -0,0 +1,61 @@
+# SUSE's openQA tests
+#
+# Copyright 2022 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+
+# Package: cloud-regionsrv-client
+# Summary: Test system (re)registration
```
Comment (grisu48) [inline tests/publiccloud/registercloudguest.pm:7]: I would add here the reference to https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/integration_test-process.txt or do you think that doesn't make sense? e.g.

    # Summary: Test system registration
    #          See https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/integration_test-process.txt for the agreed test procedure

File: tests/publiccloud/registercloudguest.pm [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+sub run {
+    my ($self, $args) = @_;
+
+    # Preserve args for post_fail_hook
+    $self->{instance} = $args->{my_instance};
+
+    # Create $instance to make the code easier to read
+    my $instance = $args->{my_instance};
+    my $regcode = (is_byos()) ? '-r ' . get_required_var('SCC_REGCODE') : '';
```
Comment (grisu48) [inline tests/publiccloud/registercloudguest.pm:28]: Same as above: A variable named `$regcode` should only contain the REGCODE and not additional variables unless strictly necessary.

File: tests/publiccloud/registercloudguest.pm [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    $instance->run_ssh_command(cmd => '! sudo SUSEConnect -d', quiet => 0) if (is_byos());
+    $instance->run_ssh_command(cmd => 'sudo registercloudguest --clean', quiet => 0);
+    $instance->run_ssh_command(cmd => '! sudo zypper lr', timeout => 600);
+    $instance->run_ssh_command(cmd => '! sudo ls /etc/zypp/credentials.d/*', quiet => 0);
+    if (is_byos()) {
+        $instance->run_ssh_command(cmd => 'sudo SUSEConnect --version');
+        $instance->run_ssh_command(cmd => "sudo SUSEConnect $regcode");
+        my $remote = $args->{my_instance}->username . '@' . $args->{my_instance}->public_ip;
+        #register_addon($remote, 'pcm');
```
Comment (grisu48) [inline tests/publiccloud/registercloudguest.pm:40]: Leftover?

File: tests/publiccloud/registercloudguest.pm [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    # Preserve args for post_fail_hook
+    $self->{instance} = $args->{my_instance};
+
+    # Create $instance to make the code easier to read
+    my $instance = $args->{my_instance};
+    my $regcode = (is_byos()) ? '-r ' . get_required_var('SCC_REGCODE') : '';
+
+    select_host_console();    # select console on the host, not the PC instance
+
+    $instance->run_ssh_command(cmd => '! sudo SUSEConnect -d', quiet => 0) if (is_byos());
```
Comment (grisu48) [inline tests/publiccloud/registercloudguest.pm:32]: Please add a comment here stating 1) why we are doing this and 2) what the pre-requirements are. e.g.

```suggestion
    # Test the re-registration. This test assumes the registration has been tested before
    $instance->run_ssh_command(cmd => '! sudo SUSEConnect -d', quiet => 0) if (is_byos());
```

**Review by grisu48 — COMMENTED** [review]
Good work Pavel! I have left some comments, but feel that we are on a good track. Keep it up!

File: lib/publiccloud/instance.pm (hunk 1 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:223]: I used `rc_only` but it did not really work for me. I'll try once more.

File: tests/publiccloud/registercloudguest.pm (hunk 9 above)
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm:40]: This is commented because of the [bsc#1198080](https://bugzilla.suse.com/show_bug.cgi?id=1198080)

Comment (asmorodskyi) [reply tests/publiccloud/registercloudguest.pm:40]: soft-failure than ?

File: lib/publiccloud/instance.pm [hunk 11]
```diff
@@ -203,10 +203,57 @@ sub wait_for_guestregister
...
+
+        if ($self->run_ssh_command(cmd => 'sudo systemctl is-enabled guestregister.service', proceed_on_failure => 1, quiet => 1) !~ /disabled/) {
```
Comment (asmorodskyi) [inline lib/publiccloud/instance.pm:225]: looks like something which needs `validate_ssh_command_output` ? if you know what I mean ?

File: tests/publiccloud/boottime.pm [hunk 12]
```diff
@@ -271,6 +271,7 @@ sub do_systemd_analyze {
 
     # Wait for guest register, before calling syastemd-analyze
     $instance->wait_for_guestregister();
+    $instance->check_guestregister();
```
Comment (asmorodskyi) [inline tests/publiccloud/boottime.pm:274]: I think this test should be exception . check one line below . following measurement does not make sense with line which you adding here ...

File: lib/publiccloud/utils.pm [hunk 13]
```diff
@@ -79,15 +79,36 @@ sub register_addon {
...
     my $remote = $instance->username . '@' . $instance->public_ip;
-    # not all images currently have registercloudguest pre-installed .
+    # not all BYOS images currently have registercloudguest pre-installed .
     # in such a case,we need to regsiter against SCC and install registercloudguest with all needed dependencies and then
     # unregister and re-register with registercloudguest
-    if ($instance->run_ssh_command(cmd => "sudo which registercloudguest > /dev/null; echo \"registercloudguest\$?\" ", proceed_on_failure => 1) =~ m/registercloudguest1/) {
-        $instance->retry_ssh_command(cmd => "sudo SUSEConnect -r $regcode", timeout => 420, retry => 3, delay => 120);
+    if (is_byos() && $instance->run_ssh_command(cmd => "sudo which registercloudguest > /dev/null; echo \"registercloudguest\$?\" ", proceed_on_failure => 1) =~ m/registercloudguest1/) {
```
Comment (asmorodskyi) [inline lib/publiccloud/utils.pm:110]: if you check current master all calls of `registercloudguest` function come with condition `if(is_byos())` or `if(!is_on_demand())` . while generally I like idea to "move" this `if` into function and don't have it "outside" . I would call this cleaner approach . But currently I see two problems : 
1.  Current PR is not "refactoring of registercloudguest function" . and the fact that you mix this change with addition of new test make it harder to review for both the test and the refactoring. It would be just easier to tackle both if they come split. And I am speaking not just two different commits but really to separate PRs 
2. While you introduce `is_byos`  internally in function you haven't drop duplicated checks which we have on every call of this function. I would say it is overkill to do this check two times .

Comment (asmorodskyi) [reply lib/publiccloud/utils.pm:110]: Regarding 1st point there is one this which could fully drop it - necessity of having them in one PR . If I overlook it please point me to it ...

File: tests/publiccloud/registercloudguest.pm (hunk 9 above)
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm:40]: The test dies in `registration::ssh_remove_suseconnect_product`. It is too complicated to softfailure there. I will remove it for now and mention it in the progress ticket.

File: lib/publiccloud/instance.pm (hunk 11 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:225]: I don't. I would first of all split and refactor `run_ssh_command`.

File: tests/publiccloud/registercloudguest.pm (hunk 8 above)
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm]: Can you please re-check now?

File: lib/publiccloud/utils.pm (hunk 13 above)
Comment (pdostal) [reply lib/publiccloud/utils.pm]: The `registercloudguest` is now as it was before. Good point, thank you

File: lib/publiccloud/instance.pm (hunk 11 above)
Comment (asmorodskyi) [reply lib/publiccloud/instance.pm]: I mean something like this https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L1245 but instead of `script_output` call you doing `run_ssh_command` . And what you have on your mind when you saying about refactor `run_ssh_command` ?

File: lib/publiccloud/instance.pm [hunk 14]
```diff
@@ -65,7 +65,7 @@ sub run_ssh_command {
 
     my $ssh_cmd = sprintf('ssh %s "%s@%s" -- %s', $args{ssh_opts}, $args{username}, $self->public_ip, $cmd);
     $ssh_cmd = "timeout $timeout $ssh_cmd" if ($timeout > 0);
-    record_info('SSH CMD', $ssh_cmd);
+    record_info('SSH CMD', $ssh_cmd) unless ($rc_only);
```
Comment (asmorodskyi) [inline lib/publiccloud/instance.pm:68]: why we don't want to show command for case when you just need return code ? We already have `quiet` option maybe better to use it here instead ?

File: lib/publiccloud/utils.pm [hunk 15]
```diff
@@ -79,6 +79,27 @@ sub register_addon {
...
 }
 
+sub deregister_addon {
+    my ($remote, $addon) = @_;
+    my $arch = get_var('PUBLIC_CLOUD_ARCH', 'x86_64');
+    $arch = "aarch64" if ($arch eq "arm64");
+    my $timestamp = utc_timestamp();
+    record_info($addon, "Going to deregister '$addon' addon\nUTC: $timestamp");
+    my $cmd_time = time();
+    # ssh_remove_suseconnect_product($remote, $name, $version, $arch, $params, $timeout, $retries, $delay)
```
Comment (asmorodskyi) [inline lib/publiccloud/utils.pm:89]: leftover ?

File: lib/registration.pm [hunk 16]
```diff
@@ -223,7 +224,7 @@ sub ssh_add_suseconnect_product {
 
     remove_suseconnect_product($name, [$version, [$arch, [$params]]]);
 
-Wrapper for SUSEConnect -d $name.
+wrapper for suseconnect -d $name.
```
Comment (asmorodskyi) [inline lib/registration.pm:227]: what the purpose of this change ?

File: lib/registration.pm [hunk 17]
```diff
@@ -233,6 +234,22 @@ sub remove_suseconnect_product {
...
 
+=head2 ssh_remove_suseconnect_product
+
+    ssh_remove_suseconnect_product($name, [$version, [$arch, [$params]]]);
+
+wrapper for suseconnect -d $name over ssh.
+=cut
+sub ssh_remove_suseconnect_product {
+    my ($remote, $name, $version, $arch, $params) = @_;
+    assert_script_run "sftp $remote:/etc/os-release /tmp/os-release";
```
Comment (asmorodskyi) [inline lib/registration.pm:245]: what about our lovely  "fingerprint have changed " error , I assume we will start getting it here in some cases ?

File: lib/registration.pm (hunk 6 above)
Comment (asmorodskyi) [reply lib/registration.pm]: this question remain unanswered , can you please clarify ?

File: tests/publiccloud/img_proof.pm [hunk 18]
```diff
@@ -112,14 +112,15 @@ sub run {
     } else {
         $provider = $self->provider_factory();
         $instance = $provider->create_instance();
+
+        $instance->wait_for_guestregister();
```
Comment (asmorodskyi) [inline tests/publiccloud/img_proof.pm:116]: by this change you fully remove `wait_for_guestregister()` for all QAM runs , don't think that this was idea , right ?
also check will be only introduced for product runs

File: lib/publiccloud/instance.pm (hunk 11 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm]: It's probably for separate discussion. Do you think that I should do this in this pull request?

File: lib/publiccloud/utils.pm (hunk 15 above)
Comment (pdostal) [reply lib/publiccloud/utils.pm:89]: No. It's valid comment. The `registration:: ssh_remove_suseconnect_product()` has too many parameters.

File: lib/registration.pm (hunk 16 above)
Comment (pdostal) [reply lib/registration.pm:227]: There is no purpose, it's a mistake caused by my new keyboard. I'm sorry

File: lib/registration.pm (hunk 17 above)
Comment (pdostal) [reply lib/registration.pm:245]: I didn't notice it lately. Plus it's in my opinion separate problem to solve.

File: tests/publiccloud/img_proof.pm (hunk 18 above)
Comment (pdostal) [reply tests/publiccloud/img_proof.pm]: Please re-check

File: lib/publiccloud/instance.pm (hunk 14 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm]: I discarded this change. It should not be needed.

File: lib/registration.pm (hunk 6 above)
Comment (pdostal) [reply lib/registration.pm]: Yes, in `vim`, if you select some text, you can press `y` to yank it or `u` to make it lowercase. I simply didn't notice I pressed `u` instead of `y`.

File: lib/publiccloud/instance.pm (hunk 11 above)
Comment (asmorodskyi) [reply lib/publiccloud/instance.pm]: +1 , this is just thoughts for the future . resolving the thread

File: lib/publiccloud/instance.pm (hunk 1 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm]: @grisu48 @asmorodskyi this is not true. See `testapi::script_output` [documentation](http://open.qa/api/testapi/#_script_output):
> Executing script inside SUT with bash -eox (in case of serial console with bash -eo) and directs stdout (not' stderr!) > to the serial console and returns the output if' the script exits with 0. Otherwise the test is set to failed.

File: lib/publiccloud/utils.pm (hunk 15 above)
Comment (asmorodskyi) [reply lib/publiccloud/utils.pm:89]: ah I see now , can you please move it two lines down so it would be clear to what it belongs

File: lib/registration.pm (hunk 17 above)
Comment (asmorodskyi) [reply lib/registration.pm:245]: ok

File: tests/publiccloud/img_proof.pm [hunk 19]
```diff
@@ -112,14 +112,15 @@ sub run {
     } else {
         $provider = $self->provider_factory();
         $instance = $provider->create_instance();
+
+        $instance->wait_for_guestregister() if is_ondemand();
```
Comment (asmorodskyi) [inline tests/publiccloud/img_proof.pm:116]: now you moved it back into "not QAM" branch , any particular reason for this ?

File: tests/publiccloud/registercloudguest.pm [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    # Create $instance to make the code easier to read
+    my $instance = $args->{my_instance};
+    my $regcode = get_required_var('SCC_REGCODE') if (is_byos());
+    my $regcode_param = ($regcode) ? "-r $regcode" : '';
+
+    select_host_console();    # select console on the host, not the PC instance
+
+    # Test re-registration. Assuming the system has been registered before
+    $instance->run_ssh_command(cmd => '! sudo SUSEConnect -d', quiet => 0) if (is_byos());
+    $instance->run_ssh_command(cmd => 'sudo registercloudguest --clean', quiet => 0);
```
Comment (asmorodskyi) [inline tests/publiccloud/registercloudguest.pm:36]: calling `registercloudguest --clean` just after `SUSEConnect -d` was called , I am not sure that this is good idea . Why you would do de-register two time in a row ?

File: tests/publiccloud/registercloudguest.pm [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+    $instance->run_ssh_command(cmd => '! sudo zypper lr', timeout => 600);
```
Comment (asmorodskyi) [inline tests/publiccloud/registercloudguest.pm:37]:
```suggestion
    $instance->run_ssh_command(cmd => 'sudo zypper lr | wc -l', timeout => 600);
```

and then checking for amount of lines is safer approach . in case of current implementation (with '!' ) you will also get green bubble in case of any errors from zypper e.g. broken repo configs

File: tests/publiccloud/registercloudguest.pm [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    $instance->run_ssh_command(cmd => '! sudo ls /etc/zypp/credentials.d/*', quiet => 0);
```
Comment (asmorodskyi) [inline tests/publiccloud/registercloudguest.pm:38]: same here , better count lines

File: tests/publiccloud/registercloudguest.pm [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+    if (is_byos()) {
+        $instance->run_ssh_command(cmd => 'sudo SUSEConnect --version');
+        $instance->run_ssh_command(cmd => "sudo SUSEConnect $regcode_param");
+        my $remote = $args->{my_instance}->username . '@' . $args->{my_instance}->public_ip;
```
[the remark this answers is not in the document]
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm:42]: We don't. It's another leftover.

File: tests/publiccloud/img_proof.pm (hunk 19 above)
Comment (pdostal) [reply tests/publiccloud/img_proof.pm:116]: Yes. In latest we create and register the instance directly in `img_proof` test module. But in QAM we have separate modules for that.

File: tests/publiccloud/registercloudguest.pm (hunk 20 above)
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm:36]: I test that `SUSEConnect -d` fails. See the [documentation](https://github.com/SUSE-Enceladus/cloud-regionsrv-client/blob/master/integration_test-process.txt):
> - SUSEConnect -d
>   + produce error message

File: tests/publiccloud/registercloudguest.pm (hunk 22 above)
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm]: Done.

File: tests/publiccloud/registercloudguest.pm (hunk 21 above)
Comment (pdostal) [reply tests/publiccloud/registercloudguest.pm]: Done.

File: lib/publiccloud/utils.pm (hunk 15 above)
Comment (pdostal) [reply lib/publiccloud/utils.pm]: Done.

**Review by asmorodskyi — APPROVED** [review]
LGTM

**Comment by asmorodskyi** [comment]
I will not merge because I am not sure if @grisu48  resolve all his concerns about this PR.

File: lib/publiccloud/instance.pm (hunk 1 above)
Comment (grisu48) [reply lib/publiccloud/instance.pm]: Thanks for the clarification, then your code is correct :+1:

File: lib/publiccloud/instance.pm [hunk 24]
```diff
@@ -203,10 +203,60 @@ sub wait_for_guestregister
...
+
+        if ($self->run_ssh_command(cmd => 'sudo ls /etc/zypp/credentials.d/ | wc -l', proceed_on_failure => 1, quiet => 1) != 0) {
+            die('/etc/zypp/credentials.d/ is not empty');
+        }
+
+        if ($self->run_ssh_command(cmd => 'sudo stat --printf="%s" /var/log/cloudregister', proceed_on_failure => 1, quiet => 1) != 0) {
+            die('/var/log/cloudregister is not empty');
+        }
+    } else {
+        if ($self->run_ssh_command(cmd => 'sudo zypper lr | wc -l', timeout => 360) < 5) {
```
Comment (grisu48) [inline lib/publiccloud/instance.pm:238]: Forgotten `quiet`? You can ignore this comment if this is only a matter of style without functional implications.

**Review by grisu48 — APPROVED** [review]
LGTM!

File: lib/publiccloud/instance.pm (hunk 24 above)
Comment (pdostal) [reply lib/publiccloud/instance.pm:238]: TBH I'd put `quiet => 0` everywhere. That would make the review easier but also the code even less readable. WDYT?

Comment (grisu48) [reply lib/publiccloud/instance.pm:238]: Yes, this is the only command where it is missing. I was wondering if that was on purpose of simply forgotten.

**Review by grisu48 — APPROVED** [review]

File: tests/publiccloud/registercloudguest.pm [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+    my ($self, $args) = @_;
+
+    # Preserve args for post_fail_hook
+    $self->{instance} = $args->{my_instance};
+
+    # Create $instance to make the code easier to read
+    my $instance = $args->{my_instance};
+
+    my $regcode = (is_byos()) ? get_required_var('SCC_REGCODE') : undef;
+    my $regcode_param = ($regcode) ? "-r $regcode" : '';
```
Comment (asmorodskyi) [inline tests/publiccloud/registercloudguest.pm:31]:
```suggestion
    my $regcode_param = (is_byos()) ? "-r " . get_required_var('SCC_REGCODE')  : '';
```

if I haven't miss something variable `$regcode` used only here , so it can be dropped

**Review by asmorodskyi — APPROVED** [review]
LGTM ( you may ignore my single comment :) )

**Comment by pdostal** [comment]
Re-verified Anton's requested change on [PAYG](http://pdostal-server.suse.cz/tests/14363) and [BYOS](http://pdostal-server.suse.cz/tests/14364).
